Rotate the encryption secrets on deploy

bennerv opened the pull request:
### Which issue this PR addresses:

[Secret rotation for database / graph / frontend keys](https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/8971701/)

### What this PR does / why we need it:

* Rotate secret keys in service keyvault on deploy

### Test plan for issue:

How did you test that this PR works?
1. deployed full RP in dev mode

### Is there any documentation that needs to be updated for this PR?

No.  

### __Additional Notes__:

* Attempting to decrypt a secret with a different encryption key of the same length will not work (an error will be returned) even for the new cipher.  This ensures that we will not get back garbage when attempting to decrypt a field with the wrong key.  
*  [Key Vault does not restrict the number of versions on a secret, key or certificate, but storing a large number of versions (500+) can impact the performance of backup operations.](https://docs.microsoft.com/en-us/azure/key-vault/general/service-limits#limits-on-count-of-keys-secrets-and-certificates)
* Requires RP-Config PR [change](https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/RP-Config/pullrequest/4851309)

bennerv commented:
This PR is branched off of #1467 so requires it to be merged first hence the /hold label.

m1kola commented on pkg/util/encryption/xchacha20poly1305.go:
```diff
@@ -53,3 +54,7 @@ func (c *xChaCha20Poly1305) Seal(input []byte) ([]byte, error) {
 }
+
+func (c *xChaCha20Poly1305) GetSealSecretVersion() string {
+	return c.secretVersion
```
where is it comming from? We do not set `secretVersion` anywhere as far as I see.

Is it supposed to be an empty string? If so, should we moreve the field and `return ""` here?

m1kola commented on pkg/util/encryption/aead.go:
```diff
@@ -6,4 +6,5 @@ package encryption
 type AEAD interface {
 	Open([]byte) ([]byte, error)
 	Seal([]byte) ([]byte, error)
+	GetSealSecretVersion() string
```
In go there is convention to not have prefixes for [getters](https://golang.org/doc/effective_go#Getters).

bennerv commented on pkg/api/openshiftcluster.go:
```diff
@@ -145,6 +145,8 @@ type OpenShiftClusterProperties struct {
 
 	RegistryProfiles []*RegistryProfile `json:"registryProfiles,omitempty"`
+
+	SecretVersion string `json:"secretVersion,omitempty"`
```
If it lives at a document level the only way to view the contents of `SecretVersion` is to query the cosmosDB directly.  The converters deal with an OCP cluster not at the document level but at the `pkg/api/openshiftcluster.go` level.  

The goal of having a `SecretVersion` is to ensure all encrypted fields within a document have been updated.  We wouldn't be able to query this in the admin API if it lives in `pkg/api/openshiftclusterdocument.go`.  

I can move it to a higher level in `pkg/api/openshiftcluster.go` instead of within `openshiftclusterproperties`.

m1kola replied:
Why do we need this in the admin API?

bennerv replied:
We can confirm all the documents are on a new secret version and remove old secrets.  Otherwise there's no way of knowing which document is encrypted with which version of the secret.

m1kola replied:
Hm. I really think that properties is not the right place. Higher layer is better, but still this doesn't feel relevant to the cluster.

If we put it on the doc level - we can create a geneva action to report on secret version. But this sounds like a bit of overkill...

bennerv replied:
Splitting the secret version out of this PR.

bennerv commented on pkg/deploy/predeploy.go:
```diff
@@ -405,6 +421,32 @@ func (d *deployer) configureServiceSecrets(ctx context.Context) error {
+	}
+
+	for _, secret := range existingSecrets {
+		if filepath.Base(*secret.ID) == secretName {
```
The current implementation of `ensureSecret` [before this change does the same](https://github.com/Azure/ARO-RP/blob/master/pkg/deploy/predeploy.go#L408-L430)

This followed suit.  I believe the reasoning is because if we just call `kv.GetSecret()`, it will return an error if the secret doesn't exist, but also will return an error if the call fails for some reason.  

Fetching all the secrets in the keyvault (`kv.GetSecrets(ctx)`) and then getting the individual data out of the secret (`kv.GetSecret(ctx, secretName, "")`) ensures that we check for the existence of the secret rather than error check the first call, which the error is subject to change over time.

bennerv commented on pkg/deploy/devconfig.go:
```diff
@@ -48,6 +48,7 @@ func deployKeyvaultAccessPolicy(_env env.Core) map[string]interface{} {
 		"objectId": os.Getenv("AZURE_SERVICE_PRINCIPAL_ID"),
 		"permissions": map[string]interface{}{
 			"secrets": []interface{}{
+				"Get",
```
no - i think the policy is already set [here](https://github.com/Azure/ARO-RP/blob/master/pkg/deploy/generator/resources_rp.go#L1211-L1212)

Good catch - double checking the testing.

bennerv replied:
Ok spent some time thinking about this now and I remember.  

This change here is intended.  We need to get individual secrets during deploy to check how long ago they were created to ensure we don't create a new secret if the secret is less than 3 days (72 hours) old.  These secrets that we're checking are in both the service keyvault and the portal keyvault shown [here](https://github.com/Azure/ARO-RP/pull/1645/files#diff-9c46df97568b8e9865152bf3f58fa6f8feb1b4d7f734c14ba7d5fc7d2d311f54R398-R400).  

This is only for the dev-config because the deployer SP needs permission to read that.  There's a matching set of updates to the Fairfax and public cloud INT/Prod instances to the RP config in this [RP Config PR](https://msazure.visualstudio.com/AzureRedHatOpenShift/_git/RP-Config/commit/3087e76f8a96b296a64cfa43e0421582ccfd4b4c?refName=refs/heads/encryption-rotation-keyvault-perms&path=/deploy/prod-config.yaml)

I did miss the additional portal access policies within the PR, but have now updated them as a result.

m1kola commented on pkg/deploy/predeploy.go:
```diff
@@ -24,6 +25,8 @@ import (
 	"github.com/Azure/ARO-RP/pkg/util/keyvault"
 )
 
+const rotateSecretAfter = time.Hour * 72
```
Do I understnad correctly that we want to rotate on each deployment? But if we had a deployment in last 72 hours - we do not rotate?

Might be worth documenting it somehow.

bennerv replied:
Yes that is correct.  In case we fail a deploy we don't keep creating new secret versions every time it fails.  

I will find a place to document.

m1kola replied:
Having a comment above the constant might be good enough

bennerv replied:
Updated.  

```
// Rotate the secret on every deploy of the RP iff the most recent
// secret is less than 3 days old
const rotateSecretAfter = time.Hour * 72

bennerv commented on pkg/util/encryption/multi.go:
```diff
@@ -5,19 +5,27 @@ package encryption
 
 func NewMulti(ctx context.Context, serviceKeyvault keyvault.Manager, secretName, legacySecretName string) (AEAD, error) {
-	key, err := serviceKeyvault.GetBase64Secret(ctx, secretName, "")
+	secret, err := serviceKeyvault.GetSecret(ctx, secretName, "")
+	if err != nil {
+		return nil, err
+	}
+
+	key, err := base64.StdEncoding.DecodeString(*secret.Value)
```
`GetBase64Secret` returns a byte slice whereas `GetSecret` returns a type of `keyvault.SecretBundle`.  We use the `SecretBundle` type to get the version of the secret using `secret.ID` hence the change.  

https://github.com/Azure/ARO-RP/pull/1645/files#diff-a4aa156e26f45940a36e5cfde25660fe265f3a4dc03302a4f42ac1406181e15fR40

bennerv replied:
Outdated - will be moved to a separate PR for this.

bennerv commented on pkg/util/encryption/aescbc.go:
```diff
@@ -50,3 +51,7 @@ func (c *aes256Sha512) Seal(input []byte) ([]byte, error) {
 }
+
+func (c *aes256Sha512) SealSecretVersion() string {
+	return c.secretVersion
```
Splitting this into a separate PR based on our conversation.

bennerv commented:
@m1kola As discussed, I removed any secret version references and only focused on secret rotation for encryption keys in this PR.  I believe I addressed feedback on all the comments now as well.

As a note, the e2e is failing because there's an issue with the Red Hat registry.  If you see this during your morning and  the Red Hat registry is available please kick off e2e again for this PR.
